Mathlib/RingTheory/LittleWedderburn.lean

```diff
@@ -0,0 +1,144 @@
+    by
+    convert this using 2
+    convert finsum_cond_eq_sum_of_cond_iff _ _
+    simp only [iff_self_iff, Set.mem_toFinset, imp_true_iff]
+  simp only [Nat.cast_sum]
+  apply Finset.dvd_sum
+  rintro ⟨x⟩ hx
+  simp only [ConjClasses.quot_mk_eq_mk]
+  set Zx := Subring.centralizer ({↑x} : Set D)
+  rw [ConjClasses.card_carrier, ←Fintype.card_congr
+        (show Zxˣ ≃* _ from unitsCentralizerEquiv _ x).toEquiv, Fintype.card_units, card_D]
+  have hZx : Zx < ⊤ := by
+    rw [lt_top_iff_ne_top]
+    intro hZx
+    simp only [Set.mem_toFinset, ConjClasses.quot_mk_eq_mk] at hx
+    refine (ConjClasses.mk_bijOn (Dˣ)).1 (Set.subset_center_units ?_) hx
+    exact Subring.centralizer_eq_top_iff_subset.mp hZx <| Set.mem_singleton _
+  letI : Field Zx := hD.field _ hZx
+  letI : Algebra Z Zx :=
+    (Subring.inclusion <| Subring.center_le_centralizer {(x : D)}).toAlgebra
+  let d := finrank Z Zx
+  have card_Zx : Fintype.card Zx = q ^ d := card_eq_pow_finrank
+  have h1qd : 1 ≤ q ^ d := by rw [← card_Zx]; exact Fintype.card_pos
+  haveI : IsScalarTower Z Zx D := ⟨fun x y z ↦ mul_assoc _ _ _⟩
+  have hdn : d ∣ n := ⟨_, (finrank_mul_finrank Z Zx D).symm⟩
+  rw [Fintype.card_units, card_Zx, Int.coe_nat_div]
+  simp only [Nat.cast_sub h1qd, Nat.cast_sub h1qn, Nat.cast_one, Nat.cast_pow]
+  suffices hd : d < n
+  · apply Int.dvd_div_of_mul_dvd
+    have aux : ∀ {k : ℕ}, ((X : ℤ[X]) ^ k - 1).eval ↑q = (q : ℤ) ^ k - 1 := by
+      simp only [eval_X, eval_one, eval_pow, eval_sub, eq_self_iff_true, forall_const]
+    rw [← aux, ← aux, ← eval_mul]
+    refine' (evalRingHom ↑q).map_dvd (X_pow_sub_one_mul_cyclotomic_dvd_X_pow_sub_one_of_dvd ℤ _)
+    exact Nat.mem_properDivisors.mpr ⟨hdn, hd⟩
+  rw [← (Nat.pow_right_strictMono hq).lt_iff_lt]
+  dsimp
+  rw [← card_D, ← card_Zx]
+  obtain ⟨b, -, hb⟩ := SetLike.exists_of_lt hZx
+  refine' Fintype.card_lt_of_injective_of_not_mem _ Subtype.val_injective (_ : b ∉ _)
+  simp only [not_exists, Set.mem_range]
+  rintro y rfl
+  exact hb y.2
+
+end InductionHyp
+
+theorem center_eq_top [Finite D] : Subring.center D = ⊤ := by
+  classical
+  cases nonempty_fintype D
+  suffices
+    ∀ (n : ℕ) (D : Type _) [DivisionRing D] [Fintype D], Fintype.card D ≤ n → Subring.center D = ⊤
+    by exact this _ D le_rfl
+  clear! D
+  intro n
+  induction' n using Nat.strong_induction_on with n IH
+  intros D _D_dr _D_fin hD
+  apply InductionHyp.center_eq_top
+  intro R hR x y hx hy
+  suffices (⟨y, hy⟩ : R) ∈ Subring.center R by exact congr_arg Subtype.val (this ⟨x, hx⟩)
+  letI R_dr : DivisionRing R := Fintype.divisionRingOfIsDomain R
+  rw [IH (Fintype.card R) _ R le_rfl]
+  · trivial
+  obtain ⟨b, -, hb⟩ := SetLike.exists_of_lt hR
+  refine' (Fintype.card_lt_of_injective_of_not_mem _ Subtype.val_injective _).trans_le hD
+  · exact b
+  simp only [not_exists, Set.mem_range]
+  rintro y rfl
+  exact hb y.2
+
+
+end LittleWedderburn
+
+def littleWedderburn (D : Type _) [hD : DivisionRing D] [Finite D] : Field D :=
```

Comment:
I guess this could be an instance in Lean 4? maybe it would be good to benchmark the affect of that somehow (if littlewedderburn isn't imported anywhere we won't see an effect elsewhere yet though).

Reply:
I tried this:) in the case of instance it should probably be fully integral domain -> field, I think

Reply:
Actually that one is potentially a little more complicated as field contains extra data (inv) that integral domain doesn't, so there is a potential for non-defeq loops.

Reply:
When you tried it what happened? Why not include it in this pr then?

Reply:
I just changed it in the code, the instance is of course not yet used so we'll have to wait and see what happens

Reply:
IsDomain + Finite -> IsField could potentially be an instance though.